Update contribution documentation for testing

SarahNadeau opened the pull request:
@k-florek and @erinyoung here's a proposal for updates to the contribution documentation so that new contributions can include testing with GitHub actions. 

The idea is that contributors will:
1. Fork the repo
2. Make a Dockerfile with a "test" stage (updated instructions & template accordingly, added an [example](example/3.12.0/Dockerfile))
2. Copy the GitHub actions workflow template and fill in for their image (updated instructions, added [template](.github/workflows/test-template.yml), and [example](.github/workflows/test-example.yml ))
3. Make sure the workflow passes, i.e. can build the test stage, in their fork
4. Submit a pull request

The example image I added should let us see how this works in practice. After this pull request, you should see an action pending your approval to run "Test example image" in the "Actions" tab. Could you let me know if this works?

FYIs:
.github/workflows/build-to-test.yml is a generic workflow to build a docker image to the "test" stage while utilizing a build cache to save time, contributors should not have to touch this.
.github/workflows/pangolin.yml - @kapsakcj I added some conditional statements to stop the workflow running (and failing) in forks. Do you have any tests for this script to make sure it's still okay?

erinyoung commented:
This looks like it's going to be really cool. 

> After this pull request, you should see an action pending your approval to run "Test example image" in the "Actions" tab.

Is there a way to see it in action in your repo?

erinyoung commented on dockerfile-template/Dockerfile:
```diff
@@ -1,32 +1,60 @@
+
+# Option 2: write your own tests in a bash script in the same directory as your Dockerfile:
+COPY my_tests.sh .
+RUN bash my_tests.sh
```
I'm a little worried about adding custom files to these dockerfiles. The way docker images build requires a certain amount of space for the end user. 

For example, spades is meant to create a de novo consensus sequence from fastq files, which are large files.

SarahNadeau replied:
If I get your concern right, multistage dockerfiles help. Any data and layers added in the "test" stage are isolated from the "app" stage. So, building to the test stage is only done on the github actions runners, the test stage image is never pushed to Docker Hub. Building to the app stage yields a nice small image to push to Docker Hub.

erinyoung replied:
That is so fancy 😵

erinyoung reviewed (commented):
Overall I think this looks pretty good. I approve of including these changes.

I think the documentation and instructions need to be fleshed out a bit, but not by only @SarahNadeau

SarahNadeau commented:
> > After this pull request, you should see an action pending your approval to run "Test example image" in the "Actions" tab.
> 
> Is there a way to see it in action in your repo?

I tried to take a screen capture and attach it here. You can also go to "Actions" and click "test example image" in this repo if you want, it won't change any code or push anything - it just runs some example tests and shows you they pass (hopefully).

https://user-images.githubusercontent.com/30396464/150606201-d5ccdc12-7b9d-49b8-97eb-dfa1c0d6b91d.mov

erinyoung commented:
Looks good to me! I vote we merge this. @k-florek or @kapsakcj , do you have any concerns?

k-florek commented:
@SarahNadeau This looks great and I am happy to merge after a fiew minor tweaks. Can you please move the new workflow templates to a new folder `.github/workflow-templates` so they have a separate location from the "production" workflows.

Also I'm not clear on the difference between `test-example.yml` and `test-template.yml`?

SarahNadeau commented:
@k-florek thanks for the feedback! I moved the template as suggested and deleted the example Dockerfile & workflow - those files were just to demo how the GitHub actions testing framework works. Perhaps #273 can be a new working example once I clean that up after this PR.

I decided to leave out the "deploy" workflow to keep this PR focused on testing alone. If you guys agree, I'll submit the deploy workflow with #273.

k-florek commented:
@SarahNadeau Awesome, thanks for all of your work on this! Adding the deploy in the #273 makes sense to me.
